Point cloud support in profile plots

nyalldawson opened the pull request:
This PR adds initial support for point cloud layers to show in profile plots.

Supported so far:

- Single color display of points
- Respects the profile curve distance tolerance setting
- Automatic refinement based on plot visible area and scale and user controllable max screen error, just like 2d point cloud renderer (works great for COPC!)
- Control over appearance of points (color, size and shape)
- Snapping to points
- 
Remaining tasks are:

- Respect 2d renderer colors for points (eg classification colors)

Also includes an option to reduce the opacity of points which are further from the profile curve

https://user-images.githubusercontent.com/1829991/165682434-d515d39d-b7d0-4172-ae9c-46e9b67daa8a.mp4

nyalldawson committed:
Add map units per pixel to QgsProfileGenerationContext

nyalldawson committed:
Add elevation profile support for point cloud layers

nyalldawson committed:
Move useful point cloud renderer enums to Qgis namespace

nyalldawson committed:
Add profile related settings to point cloud elevation properties

nyalldawson committed:
Ensure that results from final plot area are always shown in plot, even when
a job was already in progress when the plot area was changed

nyalldawson committed:
Add opacity by distance from curve effect

wonder-sk commented on src/core/elevation/qgsabstractprofilegenerator.h:
```diff
@@ -243,14 +257,39 @@ class CORE_EXPORT QgsProfileGenerationContext
+     *
+     * \see scaleFactor()
+     */
+    void setScaleFactor( double factor ) {mScaleFactor = factor;}
```
I have always disliked the "scale factor" in QgsRenderContext as it seemed very arbitrary in the rendering code that used it ("why do we need to apply scale factor here and what does it mean?"). Can it actually have other meaning than number of pixels per millimeter here? If not, I would suggest to give it a more specific name, e.g. setPixelsPerMm(...)

nyalldawson replied:
I've gone with "setDpi" / "dpi" instead :+1:

wonder-sk commented on src/core/elevation/qgsprofilerenderer.h:
```diff
@@ -108,6 +108,11 @@ class CORE_EXPORT QgsProfilePlotRenderer : public QObject
     void setContext( const QgsProfileGenerationContext &context );
 
+    /**
+     * Invalidates previous resutls from all refinable sources.
```
typo results

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.h:
```diff
@@ -0,0 +1,163 @@
+      double y;
+      double z;
+      double distance;
+      double curveDistance;
```
from a quick look it's unclear which distance means distance along the curve, and which one is the distance away from the curve? maybe a comment or renaming would help

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -0,0 +1,554 @@
+/***************************************************************************
+                         qgspointcloudlayerprofilegenerator.cpp
+                         ---------------fun!
```
a lot of fun :grin:

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -0,0 +1,554 @@
+  Q_UNUSED( point )
+  Q_UNUSED( context )
+#if 0
+  // TODO -- index
```
Is it too slow without index to be usable? but yeah, building an index on first snapPoint() would make snapping really fast :-)

nyalldawson replied:
I've implemented the index & snapping now :smile:

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -0,0 +1,554 @@
+    return false;
+
+  // this is not AT ALL thread safe, but it's what QgsPointCloudLayerRenderer does !
+  // TODO: fix when QgsPointCloudLayerRenderer is made thread safe to use same approach
```
Good point, I haven't realized the 2D renderer had this issue... Fortunately it's really just the QgsPointCloudIndex that needs to be shared among multiple threads, and the important methods are all using mutexes... but yeah, things could still break (e.g. if layer gets deleted)

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -0,0 +1,554 @@
+{
+  int nodesDrawn = 0;
+
+  // see notes about this logic in QgsPointCloudLayerRenderer::renderNodesAsync
```
You have probably based this on an earlier version of the code, recently I have simplified the logic (and made it more efficient) in QgsPointCloudLayerRenderer::renderNodesAsync() - see #48211

nyalldawson replied:
Thanks, fixed

wonder-sk commented on src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp:
```diff
@@ -0,0 +1,554 @@
+
+      res.color = mPointColor.rgba();
+
+      // TODO we may hit the limit of QVector size?
```
I hope that should be unlikely if user didn't pick extremely big tolerance for the curve and extremely low max. error... (for full-canvas rendering of point clouds we typically draw up few millions of points)

wonder-sk reviewed (approved):
Awesome work as usual :-)

nyalldawson commented on src/ui/pointcloud/qgspointcloudelevationpropertieswidgetbase.ui:
```diff
@@ -128,6 +128,136 @@
+   <item>
+    <widget class="QGroupBox" name="mLayerRenderingGroupBox">
+     <property name="title">
+      <string>Profile Chart Accuracy</string>
```
@wonder-sk something I was toying with was just reusing that setting from the 2d renderer. What do you think?

wonder-sk replied:
I thought about that too, but I am undecided. It would be nice though if user needed to adjust fewer knobs... So maybe it would make sense to default to reusing _all_ the settings from 2D renderer (point styling/color + size + shape + error), and then keep the option to override the settings (what we have now).

nyalldawson committed:
Use DPI directly instead of scale factor

nyalldawson committed:
Fix typo

nyalldawson committed:
Fix comments

nyalldawson committed:
Update remote pc node loading logic

nyalldawson committed:
Clearer variable name, small optimisation

nyalldawson committed:
Use different snap tolerances when snapping to points vs snapping
to continuous elevation surfaces

When snapping to a surface we want to be very "grabby" in the height
axis but ungrabby in the distance axis. For points we should be equally
grabby on either axis.

nyalldawson committed:
Implement snapping to point cloud points in profile

nyalldawson committed:
Minor optimisations

nyalldawson committed:
Fix warning

nyalldawson committed:
Enum class

nyalldawson committed:
Fix build

nyalldawson committed:
Update test

nyalldawson committed:
Fix build

nyalldawson commented:
Unrelated test failure
